### PR TITLE
8293833: Error mixing types with -XX:+UseCMoveUnconditionally -XX:+UseVectorCmov

### PR DESCRIPTION
**Pull request by fg1417**
After JDK-8139340, JDK-8192846 and JDK-8289422, we can vectorize
the case below by enabling -XX:+UseCMoveUnconditionally and
-XX:+UseVectorCmov:
```
// double[] a, double[] b, double[] c;
for (int i = 0; i < a.length; i++) {
    c[i] = (a[i] > b[i]) ? a[i] : b[i];
}
```

But we don't support the case like:
```
// double[] a;
// int seed;
for (int i = 0; i < a.length; i++) {
    a[i] = (i % 2 == 0) ? seed + i : seed - i;
}
```
because the IR nodes for the CMoveD in the loop is:
```
  AddI  AndI     AddD   SubD
     \  /         /     /
     CmpI        /    /
       \        /   /
      Bool     /  /
          \   / /
          CMoveD
```

and it is not our target pattern, which requires that the inputs
of Cmp node must be the same as the inputs of CMove node
as commented in CMoveKit::make_cmovevd_pack(). Because
we can't vectorize the CMoveD pack, we shouldn't vectorize
its inputs, AddD and SubD. But the current function
CMoveKit::make_cmovevd_pack() doesn't clear the unqualified
CMoveD pack from the packset. In this way, superword wrongly
vectorizes AddD and SubD. Finally, we get a scalar CMoveD
node with two vector inputs, AddVD and SubVD, which has
wrong mixing types, then the assertion fails.

To fix it, we need to remove the unvectorized CMoveD pack
from the packset and clear related map info.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293833](https://bugs.openjdk.org/browse/JDK-8293833): Error mixing types with -XX:+UseCMoveUnconditionally -XX:+UseVectorCmov

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [47ca7341](https://git.openjdk.org/jdk/pull/10627/files/47ca7341d91061ca6e1b86c2308ff046c9ac410e)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10627/head:pull/10627` \
`$ git checkout pull/10627`

Update a local copy of the PR: \
`$ git checkout pull/10627` \
`$ git pull https://git.openjdk.org/jdk pull/10627/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10627`

View PR using the GUI difftool: \
`$ git pr show -t 10627`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10627.diff">https://git.openjdk.org/jdk/pull/10627.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp [hunk 1]
```diff
@@ -1946,16 +1966,16 @@ Node_List* CMoveKit::make_cmovevd_pack(Node_List* cmovd_pk) {
...
+    return cmovd_pk;
   }
 
   if (!test_cmpd_pack(cmpd_pk, cmovd_pk)) {
     NOT_PRODUCT(if(_sw->is_trace_cmov()) {tty->print("CMoveKit::make_cmovevd_pack: cmpd pack for CmpD %d failed vectorization test", cmpd->_idx); cmpd->dump();})
-    return NULL;
+    return cmovd_pk;
   }
 
   Node_List* new_cmpd_pk = new Node_List();
```
Comment (chhagedorn) [inline src/hotspot/share/opto/superword.cpp:1981]: The following suggestion is just an idea as I was a little bit confused by how you use the return value of `make_cmovevd_pack` to remove the cmove pack and its related packs. Intuitively, I would have expected that this "make method" returns the newly created pack instead. 

Maybe it's cleaner if you split this method into a "should merge" method with the check
```
 if ((cmovd->Opcode() != Op_CMoveF && cmovd->Opcode() != Op_CMoveD) ||
      pack(cmovd) != NULL /* already in the cmov pack */) {
    return NULL;
  }
```
a "can merge" method that checks all the other constraints and an actual "make pack" method with the code starting at this line. Then you could use these methods in `merge_packs_to_cmovd` like that in pseudo-code:
```
void SuperWord::merge_packs_to_cmovd() {
  for (int i = _packset.length() - 1; i >= 0; i--) {
    Node_List* pack = _packset.at(i);
    if (_cmovev_kit.should_merge(pack)) {
      if (_cmovev_kit.can_merge(pack)) {
        _cmovev_kit.make_cmovevd_pack(pack)
      } else {
        remove_cmove_and_related_packs(pack);
      }
    }
  }
  ...
```

**Review by chhagedorn — COMMENTED** [review]
Otherwise, the fix looks reasonable to me!

File: test/hotspot/jtreg/compiler/c2/TestCondAddDeadBranch.java [hunk 2]
```diff
@@ -23,10 +23,13 @@
...
 /**
  * @test
- * @bug 8268883
+ * @bug 8268883 8293833
  * @summary C2: assert(false) failed: unscheduable graph
+ *          Error mixing types with -XX:+UseCMoveUnconditionally -XX:+UseVectorCmov
  *
  * @run main/othervm -Xcomp -XX:-TieredCompilation -XX:CompileOnly=TestCondAddDeadBranch TestCondAddDeadBranch
+ * @run main/othervm -Xcomp -XX:-TieredCompilation -XX:CompileOnly=TestCondAddDeadBranch
+ *                   -XX:+UseCMoveUnconditionally -XX:+UseVectorCmov -XX:MaxVectorSize=32  TestCondAddDeadBranch
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestCondAddDeadBranch.java:32]: As the cmove flags are C2 specific, we should also add a `@requires vm.compiler2.enabled`. Same for the other test.

**Comment by merykitty** [comment]
May I ask if we can vectorise `Bool -> Cmp` into `VectorMaskCmp` and `CMove` into `VectorBlend`, this would help vectorise the pattern you mention in the description instead of bailing out? Thanks.

File: src/hotspot/share/opto/superword.cpp (hunk 1 above)
Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp]: @chhagedorn thanks for your great suggestion. It did make the code much easier to understand. I've done the refactoring in the latest commit. Please help review. Thanks for your time!

File: test/hotspot/jtreg/compiler/c2/TestCondAddDeadBranch.java (hunk 2 above)
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/TestCondAddDeadBranch.java:32]: Done. Thanks!

**Comment by fg1417** [comment]
> May I ask if we can vectorise `Bool -> Cmp` into `VectorMaskCmp` and `CMove` into `VectorBlend`, this would help vectorise the pattern you mention in the description instead of bailing out? Thanks.

@merykitty Thanks for your kind review and question. 

That's really an interesting idea. IMO, vectorizing `Bool -> Cmp` and `CMove` separately, to support more cases, deserves a deep investigation. I'm not sure if it's feasible. But for the case in the description, even trying the idea, we still can't vectorize the case because we can't vectorize `i % 2` currently. In this way, we can't vectorize any chain involving `i % 2`.

**Review by chhagedorn — APPROVED** [review]
Thanks for doing the update, looks good to me! It's indeed much easier to follow the code now. I'll submit some testing.

File: src/hotspot/share/opto/superword.cpp [hunk 3]
```diff
@@ -1909,76 +1932,90 @@ Node* CMoveKit::is_CmpD_candidate(Node* def) const {
...
-  if (!cmovd->is_CMove()) {
-    return NULL;
-  }
-  if (cmovd->Opcode() != Op_CMoveF && cmovd->Opcode() != Op_CMoveD) {
-    return NULL;
-  }
-  if (pack(cmovd) != NULL) { // already in the cmov pack
-    return NULL;
+bool CMoveKit::is_cmove_pack_candidate(Node_List* cmove_pk) {
+  Node *cmove = cmove_pk->at(0);
```
[the remark this answers is not in the document]
Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp]: Done. Thanks.

**Comment by fg1417** [comment]
@chhagedorn thanks for your review and comments! I updated the commit to resolve the code style issue.

**Review by vnkozlov — COMMENTED** [review]
Looks reasonable.

Does `compiler/c2/irTests/TestVectorConditionalMove.java` IR test cover this case? Can you add it if it is not already?

**Review by chhagedorn — APPROVED** [review]
Thanks for the update, looks good and testing passed!

**Comment by fg1417** [comment]
> Thanks for the update, looks good and testing passed!

@chhagedorn thanks for your review and test work!

**Comment by fg1417** [comment]
> Does `compiler/c2/irTests/TestVectorConditionalMove.java` IR test cover this case? Can you add it if it is not already?

Thanks for pointing it out @vnkozlov . Updated the IR testcase in the new commit.

**Review by vnkozlov — APPROVED** [review]
Thank you for updating test.

**Comment by fg1417** [comment]
Thanks all for your review and comments. I'll integrate it.

/integrate

**Pull request closed without merging**